setup.py

```diff
@@ -275,6 +275,7 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
 ]
 mssql = [
     'pymssql~=2.1.1',
+    'pyodbc',
```

Comment:
As I comment as Slack, I think we should **REPLACE** the old mssql hook instead of and a new one. It will confuse user and we don't know which to use if want to add a new transfer between mssql and other one.

Reply:
i think it's better to be clear.
there are a number of different ways to connect to mssql in python, besides pymssql (cTDS, pyodbc, turbodbc, BCP)
i think it makes sense to be clear in the hook name.
i think that operators may use one or another hook depending on the use case.  e.g. if you are going to store to parquet you might want to use turbodbc.  or if you are bulk output to csv you might consider BCP.   executing sql maybe just pyodbc.  
these hooks don't exist now, but if we think ahead, perhaps it makes sense to plan for their eventual possible inclusion.
that's my take, anyway.

Reply:
But I think mssql is recommend using pyodbc instead of pymssql. Add they was both JDBC connector (from their introduction), so I think it's better to replace it.

Reply:
> @baolsen
> 
> > I really feel that SQL Alchemy might be useful here. Using different dialects allows us to be backwards compatible and let the user decide whether to use pymssql or pyodbc (with the extra effort that it requires).
> 
> it's an interesting idea but i think it actually increases complexity to go that way. then your hook logic, and the design of your connection object parsing, has to be able to handle arbitrary mssql libraries.
> e.g. with pyodbc we can supply arbitrary odbc connection properties as key-value in conn.extra. what would you do with these in pymssql? you could do it, but i think it is cleaner and safer to keep things separate.
> and where in the conn object would you put the dialect selector?
> you'd still have to have a extra option for odbc support at install to enable pyodbc.
> and it while this could potentially be used for switching between pyodbc and pymssql, it could not be used for turbodbc or bcp.
> 
> and i think that might be why conventionally in airflow it's the other way around -- hooks are defined in accordance with the idiosyncrasies of their connectors, and they can optionally generate a sqlalchemy connection. and this PR does provide support for that here: (
> 
> https://github.com/apache/airflow/blob/1b3b907ed6e8a45affa269b624dfe420d74424ed/airflow/providers/mssql/hooks/mssql_odbc.py#L156
> 
> )
> (the `get_uri` method is used by dbapi hook to produce sqlalchemy engine; i also provide a get_sqlalchemy connection method)

I guess the current approach makes sense. It's just a bit strange to me that DbAPIHook provides SQL Alchemy engine but then we replace it with specific libraries in the hooks that re-implement it.

The MSSQL-specific hook can implement logic particular to MSSQL+pyodbc (being the default) for now. If the user wants to change the default in future, then those specific settings wouldn't be relevant to them. The MSSQL-hook could eventually implement methods specific to other dialects. Perhaps we could validate that the dialect is one of the "supported" ones.

The dialect selector doesn't need to be in the conn object, it can be passed as an argument to the MSSQLOperator (or picked up from the extra if specified). This makes it clear in the operator signature which dialect is in use and allows the user to change it with a single line of code.

You wouldn't necessarily need an option at install for ODBC unless you made that the default. Allowing the user to specify "turbodc" as a dialect and doing a manual install of the relevant drivers / environment is no worse than the current option (which is basically, no option :) )

> @baolsen
> 
> > I really feel that SQL Alchemy might be useful here. Using different dialects allows us to be backwards compatible and let the user decide whether to use pymssql or pyodbc (with the extra effort that it requires).
> 
> it's an interesting idea but i think it actually increases complexity to go that way. then your hook logic, and the design of your connection object parsing, has to be able to handle arbitrary mssql libraries.
> e.g. with pyodbc we can supply arbitrary odbc connection properties as key-value in conn.extra. what would you do with these in pymssql? you could do it, but i think it is cleaner and safer to keep things separate.
> and where in the conn object would you put the dialect selector?
> you'd still have to have a extra option for odbc support at install to enable pyodbc.
> and it while this could potentially be used for switching between pyodbc and pymssql, it could not be used for turbodbc or bcp.
> 
> and i think that might be why conventionally in airflow it's the other way around -- hooks are defined in accordance with the idiosyncrasies of their connectors, and they can optionally generate a sqlalchemy connection. and this PR does provide support for that here: (
> 
> https://github.com/apache/airflow/blob/1b3b907ed6e8a45affa269b624dfe420d74424ed/airflow/providers/mssql/hooks/mssql_odbc.py#L156
> 
> )
> (the `get_uri` method is used by dbapi hook to produce sqlalchemy engine; i also provide a get_sqlalchemy connection method)

I suppose the approach makes sense. It just means we must use the Airflow connection object in a specific way for each DB type.

FYI with SQL Alchemy, using the DSN as the host can work if using odbc.ini files. However I'm not sure it will work with this operator since the ODBC string is being built here. Are you able to test with a DSN?

I've configured Airflow to use pyodbc for the metadata DB as follows (and am using odbc.ini with DSN=AIRFLOWDB):
mssql+pyodbc://@AIRFLOWDB

So it seems you can just pass DSN as the HOST parameter.
It also reads my Trusted_Connection=yes, from odbc.ini so I can use Kerberos to connect.

Don't know if you've seen this, perhaps easier than unpacking the ODBC-specific parameters and building the string:

from sqlalchemy.engine.url import URL
(see query parameter, allows for extra options to be passed to the underlying engine).

Not sure if it will work for you but perhaps take a look and see :)

Lastly I'd suggest nesting the ODBC parameters one level down in conn.extra rather than having them at the root. Eg conn.extra = {'odbc': {...} } rather than conn.extra = { ... }

This allows for some future-proofing. If anyone wants to add other stuff to extra, which are not meant as odbc parameters (such as create_engine_kwargs or anything else really), then its easier for them to do so without breaking the existing logic.

Reply:
> DbAPIHook provides SQL Alchemy engine but then we replace it with specific libraries in the hooks that re-implement it.

DBAPI hook does not provide an engine -- it just provides a standard way to produce the engine given a URI.  each hook may need to URI in different way.  but engine generation doesn't care about those details.

Reply:
> Are you able to test with a DSN
this hook is meant to work with DSN or driver or both

Reply:
> we must use the Airflow connection object in a specific way for each DB type
yes this seems to be the case for all hooks -- not just db hooks

Reply:
>So it seems you can just pass DSN as the HOST parameter.
>It also reads my Trusted_Connection=yes, from odbc.ini so I can use Kerberos to connect.
>Don't know if you've seen this, perhaps easier than unpacking the ODBC-specific parameters and building the string

Yes you are welcome to use DSN only.  But not everyone wants to bother defining a DSN when you can just define everything you need in airflow conn uri.  E.g. me :)

But this hook intends to support either DSN or driver + params or any reasonable combination.

Reply:
> Lastly I'd suggest nesting the ODBC parameters one level down in conn.extra rather than having them at the root. Eg conn.extra = {'odbc': {...} } rather than conn.extra = { ... }

probably a good idea

Reply:
@dstandish makes sense, thanks for explaining. Up to you if you want to nest the conn.extra stuff or not, but otherwise I have nothing to add.

Reply:
> nesting odbc params in extra

@baolsen  i do like the idea here
let me share my reservations.
here is what airflow conn uri looks like without nesting:
```
'mssql-odbc://myserver?Driver=ODBC+Driver+17+for+SQL+Server&ApplicationIntent=ReadOnly&TrustedConnection=Yes'
```
and with nesting under `odbc_keywords`:
```
'mssql-odbc://myserver?odbc_keywords=%7B%27Driver%27%3A+%27ODBC+Driver+17+for+SQL+Server%27%2C+%27ApplicationIntent%27%3A+%27ReadOnly%27%2C+%27TrustedConnection%27%3A+%27Yes%27%7D'
```

with the first form, user could create conn uri , read it, and change it pretty easily

with the second form it's not really human readable/ editable, you have to use `Connection.get_uri` to generate it.

if there was a good reason to do it i would do it but if not i prefer the readability.

with sqlalchemy, the connect kwargs, by design of dbapi, are passed directly to the method.  so there's not really a need for them in `extra`.  And i would say, for any pyodbc-specific params, i think it's cleaner if we just allow them to be passed directly to hook in initialization, like i have done with `trx_isolation`.

the tradeoff just doesn't seem worth it to dirty up the conn uri  --- what are your thoughts?

Reply:
I'm not sure why it would have an impact on the URI format.

I was thinking it could be something like how config_kwargs and session_kwargs are used in aws_hook:
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/aws_hook.py

# Dejson the 'extra' field first
extra_config = connection_object.extra_dejson
...
# Use nested {'config_kwargs;: { <dict of kwargs> } } to create AWS Config object
config = Config(**extra_config["config_kwargs"])

I took this approach because we needed to have config_kwargs and session_kwargs being stored in the extra, along with a whole bunch of other existing settings that are kept there. 

Let me know if this is roughly how you approached it. I'm surprised how it could be changing the URI, because it should just be simply un-nesting and using the same values as it would have without the nesting.

Reply:
To understand airflow's URI format see docs here: https://airflow.readthedocs.io/en/latest/howto/connection/index.html#connection-uri-format

It's the way connection objects are serialized for use in environment variables or at the command line

Reply:
Also, I am pretty sure everything that can be accomplished with pyodbc connector can be achieved with this hook as presently designed because it allows arbitrary ODBC properties to be included.  Then this exact odbc connection string is used in sqlalchemy connection string.  And you can modify other connection attributes after creating the connection.  So i think we're ok here.

Reply:
ALSO, there will never be an odbc property called `connect_kwargs`, so i don't think we have to worry about future proofing with regard to that potential addition.
But with pyodbc connect kwargs , you tend to want to reference constants defined in the library (because they are integers that have no obvious meaning) so i think this would make more sense in hook init anyway.

Reply:
Ok thought about it some more and i can add connect kwargs to hook init but don't think it makes sense in conn extra

Reply:
ok after reviewing [pyodbc docs](https://github.com/mkleehammer/pyodbc/wiki/Module) i decided to add support for connect_kwargs to both hook init and conn str.  i am not nesting the odbc properties, just making it optional to supply connect kwargs alongside them.

e.g. like so:
```
        {
          "Driver": "ODBC Driver 17 for SQL Server",
          "ApplicationIntent": "ReadOnly",
          "TrustedConnection": "Yes",
          "connect_kwargs": {
            "autocommit": "false",
            "ansi": "true",
          }
        }
```

i will probably update some of the documentation over the weekend

Reply:
That example looks awesome, very near and extensible.

A lot of newcomers struggle with the concepts of operator / hook / connection and even once those are figured out, you end up inspecting the source to try and figure out how all the connection properties are being used. 

Anyway, documentation + example is a great idea and would go a really long way to helping with that :)

(As an aside, imagine if we could somehow document a Django-style help_text on Connection, which appears on the UI somewhere, or even have an example that pops up for each hook)